octoprint_authldap/__init__.py

```diff
@@ -0,0 +1,271 @@
+# coding=utf-8
+from __future__ import absolute_import
+
+import octoprint.plugin
+from octoprint.users import UserManager, FilebasedUserManager, User
+from octoprint.settings import settings
+import ldap
+import random
+import string
+import os
+
+
+class LDAPUserManager(FilebasedUserManager, octoprint.plugin.SettingsPlugin, octoprint.plugin.TemplatePlugin):
+
+	# Login phase :
+	# - findUser called, if it return a user
+	# - checkPassword called, if it return True
+	# - login_user called with User returned by previous findUser
+
+	_localUserManager = FilebasedUserManager()
+
+	def __init__(self):
+		if settings().get(["accessControl", "userManager"]) == 'octoprint_authldap.LDAPUserManager':
+			if settings().get(["plugins", "authldap", "ldap_uri"]) is not None\
+					and settings().get(["plugins", "authldap", "ldap_bind_user"]) is not None \
+					and settings().get(["plugins", "authldap", "ldap_bind_password"]) is not None \
+					and settings().get(["plugins", "authldap", "ldap_search_base"]) is not None \
+					and settings().get(["plugins", "authldap", "ldap_query"]) is not None:
+				connection = ldap.initialize(settings().get(["plugins", "authldap", "ldap_uri"]))
+				connection.set_option(ldap.OPT_REFERRALS, 0)
+				if settings().get(["plugins", "authldap", "ldap_method"]) == 'TLS':
+					ldap_verifypeer = settings().get(
+						["plugins", "authldap", "ldap_tls_reqcert"])
+					verifypeer = ldap.OPT_X_TLS_HARD
+					if ldap_verifypeer == 'NEVER':
+						verifypeer = ldap.OPT_X_TLS_NEVER
+					elif ldap_verifypeer == 'ALLOW':
+						verifypeer = ldap.OPT_X_TLS_ALLOW
+					elif ldap_verifypeer == 'TRY':
+						verifypeer = ldap.OPT_X_TLS_TRY
+					elif ldap_verifypeer == 'DEMAND':
+						verifypeer = ldap.OPT_X_TLS_DEMAND
+					# elif ldap_verifypeer == 'HARD':
+					#   verifypeer = ldap.OPT_X_TLS_HARD
+					connection.set_option(ldap.OPT_X_TLS_REQUIRE_CERT, verifypeer)
+					try:
+						connection.start_tls_s()
+					except:
+						pass
+				try:
+					connection.simple_bind_s(settings().get(["plugins", "authldap", "ldap_bind_user"]), settings().get(["plugins", "authldap", "ldap_bind_password"]))
+					connection.unbind_s()
+
+					FilebasedUserManager.__init__(self)
+					return
+				except:
+					pass
+			settings().remove(["accessControl", "userManager"])
+			settings().remove(["plugins", "authldap", "active"])
+			settings().save()
+			os.system('kill $PPID')
+
+	def findUser(self, userid=None, apikey=None, session=None):
+		user = UserManager.findUser(self, userid=userid, session=session)
+
+		if user is not None:
```

Comment:
Just some checks for already logged in users or api access
"Copied" from FilebasedUserManager